frontend/src/__tests__/unit/testUtils/hooks.ts

```diff
@@ -0,0 +1,246 @@
+ *
+ * `renderHook` provides full control over the rendering of your hook including the ability to wrap the test component.
+ * This is usually used to add context providers from `React.createContext` for the hook to access with `useContext`.
+ * `initialProps` and props subsequently set by `rerender` will be provided to the wrapper.
+ *
+ * ```
+ * const renderResult = renderHook(({ who }: { who: string }) => useSayHello(who), { initialProps: { who: 'world' }});
+ * expectHook(renderResult).toBe('Hello world!');
+ * renderResult.rerender({ who: 'there' });
+ * expectHook(renderResult).toBe('Hello there!');
+ * ```
+ */
+export const renderHook = <
+  Result,
+  Props,
+  Q extends Queries = typeof queries,
+  Container extends Element | DocumentFragment = HTMLElement,
+  BaseElement extends Element | DocumentFragment = Container,
+>(
+  render: (initialProps: Props) => Result,
+  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
+): RenderHookResultExt<Result, Props> => {
+  let updateCount = 0;
+  let prevResult: Result | undefined;
+  let currentResult: Result | undefined;
+
+  const renderResult = renderHookRTL((props) => {
+    updateCount++;
+    prevResult = currentResult;
+    currentResult = render(props);
+    return currentResult;
+  }, options);
+
+  const renderResultExt: RenderHookResultExt<Result, Props> = {
+    ...renderResult,
+
+    isStable: () => (updateCount > 1 ? Object.is(renderResult.result.current, prevResult) : false),
+
+    isStableArray: () => {
+      // prefill return array with `false`
+      const stable: boolean[] = Array(
+        Math.max(
+          Array.isArray(prevResult) ? prevResult?.length : 0,
+          Array.isArray(renderResult.result.current) ? renderResult.result.current.length : 0,
+        ),
+      ).fill(false);
+
+      if (
+        updateCount > 1 &&
+        Array.isArray(prevResult) &&
+        Array.isArray(renderResult.result.current)
+      ) {
+        renderResult.result.current.forEach((v, i) => {
+          stable[i] = Object.is(v, (prevResult as unknown[])[i]);
+        });
+      }
+      return stable;
+    },
+
+    getUpdateCount: () => updateCount,
+
+    waitForNextUpdate: async (options) => {
+      const expected = updateCount;
+      try {
+        await waitFor(() => expect(updateCount).toBeGreaterThan(expected), options);
+      } catch {
+        throw new Error('waitForNextUpdate timed out');
+      }
+    },
+  };
+
+  return renderResultExt;
+};
+
+/**
+ * Lightweight API for testing a single hook.
+ *
+ * Prefer this method of testing over `renderHook` for simplicity.
+ *
+ * ```
+ * const renderResult = testHook(useSayHello, 'world');
+ * expectHook(renderResult).toBe('Hello world!');
+ * renderResult.rerender('there');
+ * expectHook(renderResult).toBe('Hello there!');
+ * ```
+ */
+
+export const testHook = <
+  Result,
+  Hook extends (...params: P) => Result,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  P extends any[],
```

Comment:
Curious -- `unknown` doesn't work better here? I think there are subtle differences to `unknown` and `any` in the TS infra... but we "don't know" what the params are inside the array...

Reply:
Indeed. Updated to `unknown`